File: tests/algorithms/x/Kotlin/dynamic_programming/abbreviation.kt
fun _sliceStr(s: String, start: Int, end: Int): String {
    val st = if (start < 0) 0 else start
    val en = if (end > s.length) s.length else end
    return if (st >= en) "" else s.substring(st, en)
}

fun <T> _listSet(lst: MutableList<T>, idx: Int, v: T) { while (lst.size <= idx) lst.add(v); lst[idx] = v }

var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

fun index_of(s: String, ch: String): Int {
    var i: Int = (0).toInt()
    while (i < s.length) {
        if (s[i].toString() == ch) {
            return i
        }
        i = i + 1
    }
    return 0 - 1
}

fun ord(ch: String): Int {
    var upper: String = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    var lower: String = "abcdefghijklmnopqrstuvwxyz"
    var idx: Int = (index_of(upper, ch)).toInt()
    if (idx >= 0) {
        return 65 + idx
    }
    idx = index_of(lower, ch)
    if (idx >= 0) {
        return 97 + idx
    }
    return 0
}

fun chr(n: Int): String {
    var upper: String = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    var lower: String = "abcdefghijklmnopqrstuvwxyz"
    if ((n >= 65) && (n < 91)) {
        return _sliceStr(upper, n - 65, n - 64)
    }
    if ((n >= 97) && (n < 123)) {
        return _sliceStr(lower, n - 97, n - 96)
    }
    return "?"
}

fun to_upper_char(c: String): String {
    var code: Int = (ord(c)).toInt()
    if ((code >= 97) && (code <= 122)) {
        return chr(code - 32)
    }
    return c
}

fun is_lower(c: String): Boolean {
    var code: Int = (ord(c)).toInt()
    return (((code >= 97) && (code <= 122)) as Boolean)
}

fun abbr(a: String, b: String): Boolean {
    var n: Int = (a.length).toInt()
    var m: Int = (b.length).toInt()
    var dp: MutableList<MutableList<Boolean>> = mutableListOf<MutableList<Boolean>>()
    var i: Int = (0).toInt()
    while (i <= n) {
        var row: MutableList<Boolean> = mutableListOf<Boolean>()
        var j: Int = (0).toInt()
        while (j <= m) {
            row = run { val _tmp = row.toMutableList(); _tmp.add(false); _tmp }
            j = j + 1
        }
        dp = run { val _tmp = dp.toMutableList(); _tmp.add(row); _tmp }
        i = i + 1
    }
    _listSet(dp[0]!!, 0, true)
    i = 0
    while (i < n) {
        var j: Int = (0).toInt()
        while (j <= m) {
            if ((((((dp[i]!!) as MutableList<Boolean>))[j]!!) as Boolean)) {
                if ((j < m) && (to_upper_char(a[i].toString()) == b[j].toString())) {
                    _listSet(dp[i + 1]!!, j + 1, true)
                }
                if (((is_lower(a[i].toString())) as Boolean)) {
                    _listSet(dp[i + 1]!!, j, true)
                }
            }
            j = j + 1
        }
        i = i + 1
    }
    return (((dp[n]!!) as MutableList<Boolean>))[m]!!
}

fun print_bool(b: Boolean): Unit {
    if ((b as Boolean)) {
        println(true)
    } else {
        println(false)
    }
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        print_bool(abbr("daBcd", "ABC"))
        print_bool(abbr("dBcd", "ABC"))
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}

File: tests/algorithms/transpiler/FS/dynamic_programming/abbreviation.fs
// Generated 2025-08-09 15:58 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec index_of (s: string) (ch: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable s = s
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (String.length (s)) do
            if (string (s.[i])) = ch then
                __ret <- i
                raise Return
            i <- i + 1
        __ret <- -1
        raise Return
        __ret
    with
        | Return -> __ret
let rec ord (ch: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable ch = ch
    try
        let upper: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        let lower: string = "abcdefghijklmnopqrstuvwxyz"
        let mutable idx: int = index_of (upper) (ch)
        if idx >= 0 then
            __ret <- 65 + idx
            raise Return
        idx <- index_of (lower) (ch)
        if idx >= 0 then
            __ret <- 97 + idx
            raise Return
        __ret <- 0
        raise Return
        __ret
    with
        | Return -> __ret
let rec chr (n: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable n = n
    try
        let upper: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        let lower: string = "abcdefghijklmnopqrstuvwxyz"
        if (n >= 65) && (n < 91) then
            __ret <- _substring upper (n - 65) (n - 64)
            raise Return
        if (n >= 97) && (n < 123) then
            __ret <- _substring lower (n - 97) (n - 96)
            raise Return
        __ret <- "?"
        raise Return
        __ret
    with
        | Return -> __ret
let rec to_upper_char (c: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable c = c
    try
        let code: int = ord (c)
        if (code >= 97) && (code <= 122) then
            __ret <- chr (code - 32)
            raise Return
        __ret <- c
        raise Return
        __ret
    with
        | Return -> __ret
let rec is_lower (c: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable c = c
    try
        let code: int = ord (c)
        __ret <- (code >= 97) && (code <= 122)
        raise Return
        __ret
    with
        | Return -> __ret
let rec abbr (a: string) (b: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable a = a
    let mutable b = b
    try
        let n: int = String.length (a)
        let m: int = String.length (b)
        let mutable dp: bool array array = Array.empty<bool array>
        let mutable i: int = 0
        while i <= n do
            let mutable row: bool array = Array.empty<bool>
            let mutable j: int = 0
            while j <= m do
                row <- Array.append row [|false|]
                j <- j + 1
            dp <- Array.append dp [|row|]
            i <- i + 1
        dp.[int 0].[int 0] <- true
        i <- 0
        while i < n do
            let mutable j: int = 0
            while j <= m do
                if _idx (_idx dp (int i)) (int j) then
                    if (j < m) && ((to_upper_char (string (a.[i]))) = (string (b.[j]))) then
                        dp.[int (i + 1)].[int (j + 1)] <- true
                    if is_lower (string (a.[i])) then
                        dp.[int (i + 1)].[int j] <- true
                j <- j + 1
            i <- i + 1
        __ret <- _idx (_idx dp (int n)) (int m)
        raise Return
        __ret
    with
        | Return -> __ret
let rec print_bool (b: bool) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable b = b
    try
        if b then
            printfn "%b" (true)
        else
            printfn "%b" (false)
        __ret
    with
        | Return -> __ret
print_bool (abbr ("daBcd") ("ABC"))
print_bool (abbr ("dBcd") ("ABC"))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
